examples/vision/siamese.py

```diff
@@ -0,0 +1,232 @@
+
+"""
+## Create dataset
+"""
+
+# 0 index of pairs_train contain one image of pair
+# and at index 1 the other image of pair is present
+# we can split entire vector of pairs into two havles,
+# later we can generate tf.data.Dataset using these halves
+# we will do the same for labels
+
+x_train_1 = pairs_train[:,0]
+x_train_2 = pairs_train[:,1]
+# x1.shape = (120000, 28, 28)
+
+x_test_1, x_test_2 = pairs_test[:,0],pairs_test[:,1]
+
+train_pair = tf.data.Dataset.from_tensor_slices((x_train_1, x_train_2))
+train_label = tf.data.Dataset.from_tensor_slices(labels_train)
+train_ds = tf.data.Dataset.zip((train_pair, train_label)).batch(16)
+
+test_pair = tf.data.Dataset.from_tensor_slices((x_test_1, x_test_2))
+test_label = tf.data.Dataset.from_tensor_slices(labels_test)
+test_ds = tf.data.Dataset.zip((test_pair, test_label)).batch(16)
+
+"""
+## Visualize
+"""
+
+# images to show must be integral multiple of num_col
+# if not then then they would be made so
+def visualize(dataset, to_show=6, num_col=3, predictions=None, test=False):
+
+  # handle case when user input more columns then images to show
+  num_row = to_show//num_col if to_show//num_col != 0 else 1 
+
+  # num_col must be integral multiple of to_show 
+  to_show = to_show if to_show/num_col == 0 else num_row*num_col
+
+  # plot images
+  fig, axes = plt.subplots(num_row, num_col, figsize=(5,5))
+  for images, labels in dataset.take(1):
+    for i in range(to_show):
+
+        # if row is one, axes array is one dimentional
+        if num_row == 1:
+          ax = axes[i%num_col]
+        else:
+          ax = axes[i//num_col, i%num_col]
+          
+        # images[0][i][:,:,0] -> because it is
+        # (28,28,1) and imshow takes (28,82)
+        ax.imshow(tf.concat([images[0][i],images[1][i]],axis=1), cmap='gray')
+        if test:
+          ax.set_title('True: {} | Pred: {:.5f}'.format(
+          labels[i],
+          predictions[i][0]))
+        else:
+          ax.set_title('Label: {}'.format(labels[i]))
+  if test:
+    plt.tight_layout(rect = (0,0,1.9,1.9 ), w_pad=0.0)
+  else:
+    plt.tight_layout(rect = (0,0,1.5,1.5))
+  plt.show()
+  
+visualize(train_ds)
+
+"""
+## Model
+
+There are two input layers, each lead to its own dense network which produces
+embeddings. Lambda layer will merge them using euclidean distance and merged
+layer will be fed to final network.
+"""
+
+def euclidean_distance(vects):
+    x, y = vects
+    sum_square = K.sum(K.square(x - y), axis=1, keepdims=True)
+    return K.sqrt(K.maximum(sum_square, K.epsilon()))
+
+
+
+input = Input((28,28,1))
+x = tf.keras.layers.BatchNormalization()(input)
+x = Conv2D(4, (5,5), activation = 'tanh')(x)
+x = AveragePooling2D(pool_size = (2,2))(x)
+x = Conv2D(16, (5,5), activation = 'tanh')(x)
+x = AveragePooling2D(pool_size = (2,2))(x)
+x = Flatten()(x)
+
+x = tf.keras.layers.BatchNormalization()(x)
+x = Dense(10, activation = 'tanh')(x)
+dense = Model(input, x)
```

Comment:
rename dense > tower network or something as its not a dense layer but a full network.

Reply:
I meant dense network, but sorry i should be more explicit. I refactored it as tower_network.